spec.html

```diff
@@ -22546,21 +22571,93 @@ <h1>ExecuteModule ( ) Concrete Method</h1>
       </emu-clause>
 
       <emu-clause id="sec-hostresolveimportedmodule" aoid="HostResolveImportedModule">
-        <h1>Runtime Semantics: HostResolveImportedModule ( _referencingModule_, _specifier_ )</h1>
-        <p>HostResolveImportedModule is an implementation-defined abstract operation that provides the concrete Module Record subclass instance that corresponds to the |ModuleSpecifier| String, _specifier_, occurring within the context of the module represented by the Module Record _referencingModule_.</p>
+        <h1>Runtime Semantics: HostResolveImportedModule ( _referencingScriptOrModule_, _specifier_ )</h1>
+        <p>HostResolveImportedModule is an implementation-defined abstract operation that provides the concrete Module Record subclass instance that corresponds to the |ModuleSpecifier| String, _specifier_, occurring within the context of the script or module represented by the Script Record or Module Record _referencingScriptOrModule_. _referencingScriptOrModule_ may also be *null*, if the resolution is being performed in the context of an `import()` expression, and there is no active script or module at that time.</p>
+
+        <emu-note>
+          <p>An example of when _referencingScriptOrModule_ can be *null* is in a web browser host. There, if a user clicks on a control given by</p>
+
+          <pre><code class="html">&lt;button onclick="import('./foo.mjs')">Click me&lt;/button></code></pre>
+
+          <p>there will be no active script or module at the time the `import()` expression runs. More generally, this can happen in any situation where the host pushes execution contexts with *null* ScriptOrModule components onto the execution context stack.</p>
+        </emu-note>
+
         <p>The implementation of HostResolveImportedModule must conform to the following requirements:</p>
         <ul>
           <li>
             The normal return value must be an instance of a concrete subclass of Module Record.
           </li>
           <li>
-            If a Module Record corresponding to the pair _referencingModule_, _specifier_ does not exist or cannot be created, an exception must be thrown.
+            If a Module Record corresponding to the pair _referencingScriptOrModule_, _specifier_ does not exist or cannot be created, an exception must be thrown.
+          </li>
+          <li>
+            This operation must be idempotent if it completes normally. Each time it is called with a specific _referencingScriptOrModule_, _specifier_ pair as arguments it must return the same Module Record instance.
```

Comment:
confirming; is "idempotent" the correct word here? cc @michaelficarra / #1133

Reply:
I change this line to avoid the usage of "idempotent" this way in https://github.com/tc39/ecma262/pull/1363/files. We'll need to resolve the conflict.

Reply:
Any suggestions for the wording here?

Reply:
`This operation must be deterministic with respect to its parameters if it completes normally`

Reply:
Some embedding environments may do some I/O here, e.g., to get the module, so I don't think this would be accurate.

Reply:
isn't that why "completes normally" is specifically mentioned? at the point that `(referrer, specifier)` gives a module record, that pair should always continue pointing to that module record.

Reply:
@littledan I/O is fine as long as it is observably deterministic with respect to its parameters.

Reply:
I just don't understand what the term "deterministic" would mean here. When I read that sentence, it makes me think that it's saying that the operation is a function of its parameters, which would not be the case.

Reply:
@littledan It means that if I give you the same parameters in two separate calls to that function, it should produce the same result. If you'd prefer an alternative way to state that, you should comment on #1363.

Reply:
I agree with @littledan.  "deterministic" does not have that meaning.  A deterministic function could have inputs other than its formal parameters, such as the phase of the moon, the load on the CPUI, or the load factor on a cache.  It may well be "deterministic" what will happen in all those cases but that doesn't guarantee that successive calls with the specific argument will return the same result.

Reply:
@allenwb That's why the phrase is "deterministic with respect to", to tell you what are the inputs for that determinism.

Reply:
Let's move this discussion to #1363. I'll update this patch when we it lands.